### DISCUSSION TITLE
source code - _ _ init _ .py  - restructure or not ?

### DISCUSSION THREAD

**Discussion by infradom** [wills106/homeassistant-solax-modbus#70] [Ideas]
This is just an idea, not sure whether it is feasisble or not. I launch this topic just to get feedback.
As more and more models are added, the sequential reading of registers becomes messy with a lot of if statements.

Maybe we can find a way to determine the sequence automatically. This would involve following changes:

- Give each sensor declaration in const.py a corresponding register addreess
- Let a new simple algorithm analyze and sort all required registers for a certain inverter, and let it perform groups of (nearly) adjacent register reads. The results are stored in a intermediate python dictionary.
- Some registers need some processing/normalization, for these registers there is a possibility to add some customization code

In short, the if statements in  _ _ init _ _.py would disappear, and the sequence would not matter anymore.
I am willing to try to make this change, but there is one disadvantage:

- the new central sorting algorithm becomes a fairly strange generic piece of code. It should be stable and independent of the inverter types, but if something goes wrong, it may be more difficult to debug.

Please shoot ...

**Comment by wills106** [comment]
I think we need to wait and see how the X3 MIC works.
Looking at the X3 MIC Pro the registers are in a completely different location.

So we might need to perform two serial number checks.

Potentially holding register 0x0 to 0x2ff is reserved and then the serial number is at 0x300 - 0x306
The input registers are also different they start at 0x400 which is PV1 Power.

So I don't know yet if the X3 MIC is the same, or if it uses conventional Hybrid style registers.
